src/signature.c

```diff
@@ -699,17 +695,25 @@ static CK_RV p11prov_sig_op_init(void *ctx, void *provkey, CK_FLAGS operation,
         break;
     case CKF_VERIFY:
         if (class != CKO_PUBLIC_KEY) {
-            return CKR_KEY_TYPE_INCONSISTENT;
+            key = p11prov_get_pub(key);
+            if (key == NULL) {
+                return CKR_KEY_TYPE_INCONSISTENT;
+            }
         }
         break;
     default:
         return CKR_GENERAL_ERROR;
     }
+    sigctx->key = p11prov_obj_ref(key);
+    if (sigctx->key == NULL) {
+        return CKR_KEY_NEEDED;
+    }
     sigctx->operation = operation;
 
     if (digest) {
         ret = p11prov_digest_get_by_name(digest, &sigctx->digest);
         if (ret != CKR_OK) {
+            p11prov_obj_free(sigctx->key);
```

Comment:
This is wrong, if you free the key here you also need to NULL the pointer or we will get a double free in p11prov_sig_freectx(), but you do not have to null anything here exactly because p11prov_sig_freectx(0 already takes care of this.

Reply:
If you are concerned about someone potentially calling the init operation multiple times, just check if the sigctx->key is not NULL at the start of the function and immediately error if that is the case. But there is no need to add that, it is an application error to try to initialize multiple times and the worst case is some memory leakage.